src/Api/Upload.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Kirby\Api;
+
+use Kirby\Cms\App;
+use Kirby\Filesystem\Dir;
+use Kirby\Filesystem\F;
+use Kirby\Toolkit\Str;
+
+/**
+ * The Upload class supports file uploads in the
+ * context of the API
+ *
+ * @package   Kirby Api
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ * @since     4.3.0
+ */
+class Upload
+{
+	/**
+	 * Handle chunked uploads by merging all chunks
+	 * in the tmp directory and only returning the new
+	 * $source path to the tmp file once complete
+	 */
+	public static function chunk(
+		Api $api,
+		string $source,
+		string $filename
+	): string|null {
+		// if the file is uploaded in chunks
+		if ($isChunk = $api->requestHeaders('X-Upload-Chunk')) {
```

Comment:
The `X-` prefix for custom headers is [deprecated](https://datatracker.ietf.org/doc/html/rfc6648). I went looking for how the tus protocol does it and we could use it as inspiration:

- `Upload-Length` would carry the total file size. If present, we know that we need to engage chunk mode.
- `Upload-Offset` would carry the byte offset where the current chunk begins. We could use this for additional protections:
  - If the offset is 0 but the file with the provided unique ID already exists, fail (could occur if two clients generate the same random ID for whatever reason).
  - If the offset is non-0, compare it with the current file size. If they differ, we got the requests out of order, so fail.
- `Upload-Id` would carry the unique ID from the client.

I know this increases complexity, but I think it could well be worth it for added robustness.

Reply:
How do we know which is the last chunk? Do we then have to measure the size of each chunk on the backend again?

Reply:
If `Upload-Length === Upload-Offset + size of current chunk`, the file is complete. We could implement this by first appending the current chunk and then comparing the file size on disk with `Upload-Length`. If they are the same, we are done.

Reply:
Somehow, the sizes don't add up for me at the end:
987,268,918 is the `Upload-Length`, but the last chunk also sends an offset of 980,208,771 and has a length of 7,060,476 which is 987,269,247 which is larger than the whole file should be.

Reply:
https://github.com/getkirby/kirby/pull/6421/commits/bebeea708c95ac25b39decd10755b9398112f5b4#diff-3cd311828c9ff4560c3674cf73afee2c3e0d7d16a866e80630cde300c47a9aecR58-R60 also still fails for the 2nd chunk as `F::size()` returns 0.

Reply:
Maybe we need `clearstatcache()` before checking the file sizes, especially when something was written to the file directly before. If it still doesn't line up, something is wrong either client-side or server-side, potentially producing corrupt files.

Reply:
Referenced the wrong variable in a path 🙈

Reply:
I think I covered all suggestions here @lukasbestle, but please have a look yourself and mark as resolved if you agree.